### PR TITLE
New: `vue/html-indent` rule (fixes #46)

### LINKED ISSUE #46: Rule proposition: Template Indentation
It would be great to have the ability of having only one style for the template/render function HTML, by enforcing an indentation style.

It could be configured by setting the option to `tab` for tabs or a positive number for spaces.

**Proposal:**
1. For tabs `'vue/indent': 'tab'`
```
<application>
	<component />
</application>
```

2. For 2 spaces: `'vue/indent': 2`
```
<application>
  <component />
</application>
```

2. No indentation: `'vue/indent': 0`
```
<application>
<component />
</application>
```

### PR DESCRIPTION
**Pull request by mysticatea**
Fixes #46.

This PR adds `vue/html-indent` rule.

The strategy of the `vue/html-indent` rule is the same as core `indent` rule. I.e., it makes the base token and offset for every token in the first traversing, then at leaving the root element, it calculate and validate expected indentation for each line.

Currently, about inside of directives and mustaches, this `vue/html-indent` rule supports only the syntax ECMAScript 2017 standard includes. It just ignores other syntaxes. The indentation rule is [one of most complex rules](https://github.com/eslint/eslint/blob/bd09cd50283760b57949802c55a6e2b5862c6d15/lib/rules/indent.js). I think this is a good start point.

### CODE REVIEW

**Comment by mysticatea** [comment]
Thank you for your patience. I think I have done. Please review and suggest lacking tests!

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,3 +1,4 @@
 {
-    "editor.tabSize": 2
+    "editor.tabSize": 2,
```
Comment (Aladdin-ADD) [inline .vscode/settings.json]: seems not related to the `vue/html-indent`. I think it can be in another PR?

File: package.json [hunk 2]
```diff
@@ -10,7 +10,8 @@
     "lint": "eslint .",
     "pretest": "npm run lint",
     "preversion": "npm run update && npm test",
-    "update": "node ./tools/update-rules.js"
+    "update": "node ./tools/update-rules.js",
+    "watch": "warun lib/rules/html-indent.js tests/lib/rules/html-indent.js -- nyc --reporter lcov -- mocha tests/lib/rules/html-indent.js"
```
Comment (Aladdin-ADD) [inline package.json]: what is the `warun`? I didn't see it in deps.

File: lib/rules/html-indent.js [hunk 3]
```diff
@@ -0,0 +1,1228 @@
...
+
+  /**
+   * Get the first and last tokens of the given node.
+   * If the node is parenthesized, this gets the outermost parentheses.
+   * @param {Node} node The node to get.
+   * @param {number} [borderOffset] The least offset of the first token. Defailt is 0. This value is used to prevent false positive in the following case: `(a) => {}` The parentheses are enclosing the whole parameter part rather than the first parameter, but this offset parameter is needed to distinguish.
+   * @returns {{firstToken:Token,lastToken:Token}} The gotten tokens.
+   */
+  function getFirstAndLastTokens (node, borderOffset) {
+    borderOffset |= 0
```
Comment (Aladdin-ADD) [inline lib/rules/html-indent.js]: could you explain a little? :)

Comment (armano2) [reply lib/rules/html-indent.js]: its js `magic` assign 0 if value is non numeric

Comment (Aladdin-ADD) [reply lib/rules/html-indent.js]: hah.. got it! amazing! lol~

File: package.json (hunk 2 above)
Comment (mysticatea) [reply package.json]: Oops, it was added unintentionally. I was using it to test.
Thank you.

File: lib/rules/html-indent.js (hunk 3 above)
Comment (mysticatea) [reply lib/rules/html-indent.js]: Yes, it is "casting to signed integer".

Bitwise operators (`&`, `|`, "^", `~`, `<<`, `>>`, and `>>>`) casts both operands to 32bits signed integer, then the logical OR with `0` does nothing. As the result, the `x | 0` expression casts `x` to 32bits signed integer.

Comment (armano2) [reply lib/rules/html-indent.js]: @mysticatea like i said `magic`

File: tests/lib/rules/html-indent.js [hunk 4]
```diff
@@ -0,0 +1,4214 @@
...
+      </template>
+    `,
+
+    // VStartTag
+    unIndent`
+      <template>
+          <div
+              aaa
+              bbb
+    `,
```
Comment (Aladdin-ADD) [inline tests/lib/rules/html-indent.js]: not a valid template, is is intentional?

and why the parser not reporting an error?

File: docs/rules/html-indent.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+```
+
+## :wrench: Options
+
+```json
+{
+    "vue/html-indent": ["error", type, {
+        "attribute": 1,
+        "closeBracket": 0,
+        "switchCase": 0,
```
Comment (Aladdin-ADD) [inline docs/rules/html-indent.md]: could you add an example for the `switchCase` option?

File: tests/lib/rules/html-indent.js [hunk 6]
```diff
@@ -0,0 +1,4214 @@
...
+              :other-attr="{
+                  aaa: 1,
+                  bbb: 2
+              }"
+              @other-attr2="
+                  foo();
+                  bar();
+              "
+          >
+              {{
```
Comment (Aladdin-ADD) [inline tests/lib/rules/html-indent.js]: maybe we need some more tests for `{{}}`.

and I didn't see tests for comma, e.g:
```html
<template>
{{
  foo,
  bar,
  qux
}}
</template>
```

File: lib/rules/html-indent.js [hunk 7]
```diff
@@ -0,0 +1,1228 @@
...
+  function validate (tokens, comments) {
+    // Calculate and save expected indentation.
+    const firstToken = tokens[0]
+    const actualIndent = firstToken.loc.start.column
+    const expectedIndent = getExpectedIndent(tokens)
+    if (expectedIndent === Number.MAX_SAFE_INTEGER) {
+      return
+    }
+
+    // Debug log
```
[the remark this answers is not in the document]
Comment (mysticatea) [reply lib/rules/html-indent.js]: Yeah!
But I'd like to leave it to investigate for bug reports, since the debug code is long...

File: tests/lib/rules/html-indent.js (hunk 4 above)
Comment (mysticatea) [reply tests/lib/rules/html-indent.js]: Yes, it's intentional. The HTML spec includes recovering process for invalid syntax, so `vue-eslint-parser` reports syntax errors with the `programNode.templateBody.errors` property instead of throwing the error. Then [vue/no-parsing-error](https://github.com/vuejs/eslint-plugin-vue/blob/94f387ed68149f8d12c1960d68cb5d9c376b9a89/docs/rules/no-parsing-error.md) rule would report those syntax errors.

The purpose of this test case is to check whether `vue/html-indent` rule works properly on an incomplete code. Especially, the expected indentation of closing brackets (`>`) does not apply to the last attribute even if the closing bracket is lacking.

File: docs/rules/html-indent.md (hunk 5 above)
Comment (mysticatea) [reply docs/rules/html-indent.md]: Or maybe I should remove the option from this rule since `switch` is [a prohibited keyword](https://github.com/vuejs/vue/blob/c0da43d22f8f4b9aeb4f49d4e86cd9704daaff3f/src/compiler/error-detector.js#L7-L11).

File: tests/lib/rules/html-indent.js (hunk 6 above)
Comment (mysticatea) [reply tests/lib/rules/html-indent.js]: In technically, mustaches (`{{foo}}`) and directive values (`"foo"` of directives) are completely same for indentation. Both are `VExpressionContainer`. So I don't think we need many tests for mustaches. I added some tests.

File: lib/rules/html-indent.js [hunk 8]
```diff
@@ -0,0 +1,1228 @@
...
+        const node = nodeList[i]
+        if (node == null) {
+          // Holes of an array.
+          continue
+        }
+        const elementTokens = getFirstAndLastTokens(node, lastToken != null ? lastToken.range[1] : 0)
+
+        // Collect related tokens.
+        // Commas between this and the previous, and the first token of this node.
+        if (lastToken != null) {
```
[the remark this answers is not in the document]
Comment (mysticatea) [reply lib/rules/html-indent.js]: I apology that I'm inactive during a long time.

It, ensuring boolean type for conditional expression, is just my favorite style :smile:
